src/Psalm/Internal/Analyzer/Statements/Expression/AssertionFinder.php

```diff
@@ -1499,50 +1501,35 @@ protected static function hasNonEmptyCountEqualityCheck(
         PhpParser\Node\Expr\BinaryOp $conditional,
         ?int &$min_count
     ) {
-        $left_count = $conditional->left instanceof PhpParser\Node\Expr\FuncCall
+        if ($conditional->left instanceof PhpParser\Node\Expr\FuncCall
             && $conditional->left->name instanceof PhpParser\Node\Name
             && strtolower($conditional->left->name->parts[0]) === 'count'
-            && $conditional->left->getArgs();
-
-        $operator_greater_than_or_equal =
-            $conditional instanceof PhpParser\Node\Expr\BinaryOp\Greater
-            || $conditional instanceof PhpParser\Node\Expr\BinaryOp\GreaterOrEqual;
-
-        if ($left_count
-            && $conditional->right instanceof PhpParser\Node\Scalar\LNumber
-            && $operator_greater_than_or_equal
-            && $conditional->right->value >= (
-                $conditional instanceof PhpParser\Node\Expr\BinaryOp\Greater
-                ? 0
-                : 1
-            )
+            && $conditional->left->getArgs()
+            && ($conditional instanceof BinaryOp\Greater || $conditional instanceof BinaryOp\GreaterOrEqual)
         ) {
-            $min_count = $conditional->right->value +
-                ($conditional instanceof PhpParser\Node\Expr\BinaryOp\Greater ? 1 : 0);
-
-            return self::ASSIGNMENT_TO_RIGHT;
-        }
-
-        $right_count = $conditional->right instanceof PhpParser\Node\Expr\FuncCall
+            $assignment_to = self::ASSIGNMENT_TO_RIGHT;
+            $compare_to = $conditional->right;
+            $comparison_adjustment = $conditional instanceof BinaryOp\Greater ? 1 : 0;
+        } elseif ($conditional->right instanceof PhpParser\Node\Expr\FuncCall
             && $conditional->right->name instanceof PhpParser\Node\Name
             && strtolower($conditional->right->name->parts[0]) === 'count'
-            && $conditional->right->getArgs();
-
-        $operator_less_than_or_equal =
-            $conditional instanceof PhpParser\Node\Expr\BinaryOp\Smaller
-            || $conditional instanceof PhpParser\Node\Expr\BinaryOp\SmallerOrEqual;
+            && $conditional->right->getArgs()
+            && ($conditional instanceof BinaryOp\Smaller || $conditional instanceof BinaryOp\SmallerOrEqual)
+        ) {
+            $assignment_to = self::ASSIGNMENT_TO_LEFT;
+            $compare_to = $conditional->left;
+            $comparison_adjustment = $conditional instanceof BinaryOp\Smaller ? 1 : 0;
+        } else {
+            return false;
+        }
 
-        if ($right_count
-            && $conditional->left instanceof PhpParser\Node\Scalar\LNumber
-            && $operator_less_than_or_equal
-            && $conditional->left->value >= (
-                $conditional instanceof PhpParser\Node\Expr\BinaryOp\Smaller ? 0 : 1
-            )
+        // TODO get node type provider here somehow and check literal ints and int ranges
```

Comment:
you can pass $source from caller method and then do `$source->getNodeTypeProvider()` if you want

Reply:
I can't believe I missed that, I was looking for a `StatementsAnalyzer`... I'm about done for now, but I'll come fix that in the next day or so.

Reply:
otherwise, the rest seems good :) Give me a sign when it's ready ;)

Reply:
Ok, well this is actually a bit complicated. I got it all working, but I got a couple `RedundantCondition`s... I have to figure out a way to make the assertion not redundant if `$foo` and `$bar` are both non-empty and there's an assertion for `count($foo) === count($bar)`.

Go ahead and merge this, I'll have to dig deeper later. I'll probably see if I can fix this on master along with #7096.